python_ta/typecheck/base.py

```diff
@@ -219,207 +258,203 @@ def fresh_tvar(self, node: Optional[NodeNG] = None) -> TypeVar:
+            node.parent = node
         return node
 
+    def get_tnode(self, t: type, add_to_nodes=True) -> _TNode:
+        try:
+            node = self.type_to_tnode[str(t)]
+            return node
+        except KeyError:
+            node = self._make_set(t, None, add_to_nodes)
+            return node
+
     ###########################################################################
     # Type lookup ("find")
     ###########################################################################
-    def resolve(self, t: type) -> TypeResult:
-        """Return the type associated with the given type.
-
-        If t is a type variable that is associated with a concrete (non-TypeVar) type, return the concrete type.
-        Otherwise if the type variable with the smallest name is returned (using < to compare strings).
+    def resolve(self, t: type) -> TypeInfo:
+        """Return the concrete type associated with the given type.
         """
-        if isinstance(t, TypeVar):
-            return TypeInfo(self._find(t).type)
-        elif isinstance(t, GenericMeta):
-            return self._resolve_generic(t)
-        else:
-            return TypeInfo(t)
-
-    def _resolve_generic(self, t: GenericMeta) -> TypeResult:
-        # TODO: Fix duplicate code (see _unify_generic)
-        resolve_result = failable_collect([self.resolve(arg) for arg in t.__args__])
-        if isinstance(resolve_result, TypeFail):
-            return resolve_result
-
-        resolved_args = resolve_result.getValue()
-
-        gorg = _gorg(t)
-        if gorg == Tuple:
-            tuple_args = tuple(resolved_args)
-            # Handle the special case when t1 or t2 are empty tuples; TODO: investigate this
-            if tuple_args == ((),):
-                tuple_args = ()
-            return TypeInfo(Tuple[tuple_args])
-        elif gorg == Callable:
-            return TypeInfo(Callable[resolved_args[:-1], resolved_args[-1]])
-        else:
-            return TypeInfo(gorg[tuple(resolved_args)])
-
-    def _find(self, tv: TypeVar) -> _TNode:
-        """Return the disjoint set node associated with the given type variable."""
-        node = self._tvar_to_tnode[tv]
-        while node.parent is not None or (node.parent and node != node.parent):
-            node = node.parent
-        return node
+        if isinstance(t, GenericMeta):
+            res_args = [self.resolve(arg).getValue() for arg in t.__args__]
+            return _wrap_generic_meta(_gorg(t), res_args)
+        elif isinstance(t, TypeVar):
+            try:
+                par = self.find_parent(self.type_to_tnode[str(t)])
+                if par:
+                    return self.resolve(par.type)
+            except KeyError:
+                return TypeInfo(t)
+        return TypeInfo(t)
+
+    def is_concrete(self, type):
+        if isinstance(type, GenericMeta):
+            return all([self.is_concrete(arg) for arg in type.__args__])
+        return not isinstance(type, TypeVar)
```

Comment:
In general prefer putting the `return` inside an `else` branch rather than below.